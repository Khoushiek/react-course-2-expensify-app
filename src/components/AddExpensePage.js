import React from 'react';
import ExpenseFormPage from "./ExpenseForm";
import {connect} from "react-redux";
import {addExpense} from "../action/expenses";

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.addExpense(expense);
    this.props.history.push('/');
  }
  render (){
     return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseFormPage
          onSubmit = {this.onSubmit} 
        />
      </div>
     );
  };
};

const mapDispatchToProps = (dispatch) => ({
  addExpense: (expense) => dispatch(addExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);