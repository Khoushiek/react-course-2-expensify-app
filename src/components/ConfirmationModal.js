import React from 'react';
import Modal from 'react-modal';

const ConfirmationModal = (props) => (
  <Modal
    isOpen = {props.isModalOpen}
    contentLabel="Remove Expense Modal"
    onRequestClose={props.handleModalClose}
    closeTimeoutMS={200}
    className="modal"
  >
    <h3 className="modal__title">Do you really want to remove the expense?</h3>
    <div className="btn-grp">
      <button className="box-layout__button button--confimation" onClick={props.onRemoveExpense}>
        Confirm
      </button>
      <button className="box-layout__button button--confimation" onClick={props.handleModalClose}>
        Cancel
      </button>
    </div>
  </Modal>
);

export default ConfirmationModal;
