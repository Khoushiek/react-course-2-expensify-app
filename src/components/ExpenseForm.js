import React from "react";
import moment from "moment";
import {SingleDatePicker} from "react-dates";

export default class ExpenseForm extends React.Component {
  
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : "",
      amount: props.expense ? (props.expense.amount/100).toString() : "",
      note: props.expense ? props.expense.note : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ""
    }
  };
  
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({description}))
  };

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({note}))
  };

  onAmountChange = (e) => 
  {
    const amount = e.target.value;
    if (amount.match(/^\d{1,}(\.\d{0,2})?$/))
    {
      this.setState(() => ({amount}));
    }
  };

  onDateChange = (createdAt) => {
    if (createdAt){
      this.setState(() => ({createdAt}))
    }
  };

  oncalendarFocusChange = ({focused}) => {
    this.setState(() => ({calendarFocused: focused}))
  };

  onSubmit =(e) => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({error: "Please Provide a Description and Amount"}))
    }
    else {
      this.setState(() => ({error: ""}))
      this.props.onSubmit({
        description: this.state.description,
        createdAt: this.state.createdAt.valueOf(),
        amount: parseFloat(this.state.amount)*100,
        note: this.state.note
      });
    }
  }

  render(){
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form-error">{this.state.error}</p>}
        <input 
            value={this.state.description}
            className="text-input"
            type="text"
            autoFocus
            placeholder="Description"
            onChange={
              this.onDescriptionChange
            }
          />
          <input
            value={this.state.amount}
            className="text-input"
            type="text"
            placeholder="Amount"
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.oncalendarFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => (false)}
          />
          <textarea
            value={this.state.note}
            placeholder="Add a Expense note"
            className="text-area"
            onChange={this.onNoteChange}
          />
          <div>
            <button className="box-layout__button">Save Expense</button>
          </div>
      </form>
    )
  }
};