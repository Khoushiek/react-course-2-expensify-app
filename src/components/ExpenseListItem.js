import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";

const numberFormat = (value) =>
  new Intl.NumberFormat('en-IN', {
    style: 'currency',
    currency: 'INR'
  }).format(value);

export default (props) => (
  <div>
    <Link to={`/edit/${props.id}`}>
      <h3>{props.description}</h3>
    </Link>
    <p>
      {numberFormat(props.amount / 100 )}
      &nbsp; - &nbsp;  
      {moment(props.createdAt).format("MMMM Do YYYY")}</p>
  </div>
);
