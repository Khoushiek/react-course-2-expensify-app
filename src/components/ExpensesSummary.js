import React from "react";
import selectExpenses from "../selectors/expenses";
import totalExpenses from "../selectors/expenses-total";
import { connect } from "react-redux";

const numberFormat = (value) =>
  new Intl.NumberFormat('en-IN', {
    style: 'currency',
    currency: 'INR'
  }).format(value);


export const ExpensesSummary = (props) => {
  const expensesTotal = totalExpenses(props.expenses);
  const expensesCount = props.expenses.length
  return (
    <div>
      {expensesCount === 1
         ? <h3>You are viewing {expensesCount} expense totalling {numberFormat(expensesTotal / 100 )}</h3>
         : <h3>You are viewing {expensesCount} expenses totalling {numberFormat(expensesTotal / 100 )}</h3>}
    </div>
  );
};

const mapPropToState = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

export default connect(mapPropToState)(ExpensesSummary);
