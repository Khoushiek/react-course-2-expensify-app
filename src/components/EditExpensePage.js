import React from 'react';
import {connect} from "react-redux";
import ExpenseForm from "./ExpenseForm";
import {startEditExpense, startRemoveExpense} from "../action/expenses";
import ConfirmationModal from "./ConfirmationModal";

export class EditExpensePage extends React.Component {

  constructor(props){
    super(props);
    this.state = {
      isModalOpen: false
    };
  }

  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense)
    this.props.history.push("/")
  };

  onRemoveExpense = () => {
    this.props.startRemoveExpense({id: this.props.expense.id})
    this.props.history.push("/")
    this.setState(() => ({
      isModalOpen: false
    }));
  }

  handleModalOpen = () => {
    this.setState(() => ({
      isModalOpen: true
    }));
  }

  handleModalClose = () => {
    this.setState(() => ({
      isModalOpen: false
    }));
  };

  render() {
    const disableClass = this.state.isModalOpen ? 'disabled' : '';
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h2>Edit Expense</h2>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
          expense = {this.props.expense}
          onSubmit = {this.onSubmit}
          disableClass={disableClass}
          />
          <button className="box-layout__button button--secondary" onClick= {this.handleModalOpen} disabled={disableClass}>Remove Expense</button>
        </div>
        <ConfirmationModal onRemoveExpense={this.onRemoveExpense} isModalOpen={this.state.isModalOpen} handleModalClose={this.handleModalClose}/>
      </div>
    );
  }
};

const mapStateToProp = (state, props) => ({ 
  expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

const mapDispatchToProp = (dispatch, props) => (
  {
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
  } 
);

export default connect(mapStateToProp, mapDispatchToProp)(EditExpensePage);