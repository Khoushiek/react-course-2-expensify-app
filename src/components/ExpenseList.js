import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";
import ExpenseListFilter from "./ExpenseListFilter";

export const ExpenseList = (props) => (
  <div>
    <ExpenseListFilter />
    {props.expenses.length === 0 
      ? <p>No Expenses</p>
      : <h3>Expenses List</h3>}
    {props.expenses.map((expense, index) => {
      return <ExpenseListItem 
      id={expense.id}
      key = {expense.id} 
      description={expense.description} 
      amount={expense.amount} 
      createdAt={expense.createdAt}
      />;
    })}
  </div>
);

const mapPropToState = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  }
};

export default connect(mapPropToState)(ExpenseList);