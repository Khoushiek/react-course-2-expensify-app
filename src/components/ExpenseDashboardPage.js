import React from 'react';
import ExpenseList from "./ExpenseList";
import ExpensesSummary from "./ExpensesSummary";

const ExpenseDashboardPage = () => (
  <div>
    <ExpensesSummary />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;