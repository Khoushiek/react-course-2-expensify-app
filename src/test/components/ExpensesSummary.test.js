import {shallow} from "enzyme";
import expenses from "../fixtures/expenses";
import React from "react";
import {ExpensesSummary} from "../../components/ExpensesSummary";

test("Should return a expenses summary for single expense", () => {
  const wrapper = shallow(<ExpensesSummary 
                           expenses={[expenses[0]]}
                            />);
  expect(wrapper).toMatchSnapshot();
});

test("Should return a expenses summary for multiple expense", () => {
  const wrapper = shallow(<ExpensesSummary 
                           expenses={expenses}
                            />);
  expect(wrapper).toMatchSnapshot();
})