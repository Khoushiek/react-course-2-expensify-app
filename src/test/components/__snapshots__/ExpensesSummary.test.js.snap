// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should return a expenses summary for multiple expense 1`] = `
<div>
  <h3>
    You are viewing 
    3
     expenses totalling 
    ₹292.50
  </h3>
</div>
`;

exports[`Should return a expenses summary for single expense 1`] = `
<div>
  <h3>
    You are viewing 
    1
     expense totalling 
    ₹250.00
  </h3>
</div>
`;
